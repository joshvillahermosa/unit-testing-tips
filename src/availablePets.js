
/**
 * @function returnAvailablePets
 * @param {[Object]} pets Array of Pets. Structure follows: { ..., code: Number, vaccinated: Boolean, ... }
 * @param {Object} CODES Object of Codes. Structure follows: { ..., AVAILABLE: Number, ... }
 * @return {[Object]} Array of available pet
 */
const returnAvailablePets = (pets, CODES) => {
  let response = { success: false, details: '' }

  if (!Array.isArray(pets)) {
    response.details = 'pets passed is not type array'
    return response
  }

  if (pets.length < 1) {
    response.details = 'pets passed in does not have any items'
    return response
  }

  const petsAvailable = pet => pet.code === CODES.AVAILABLE && pet.vaccinated

  return pets.filter(petsAvailable)
}

export { returnAvailablePets }