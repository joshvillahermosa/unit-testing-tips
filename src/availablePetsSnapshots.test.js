/**
 * @fileoverview
 * Example of testing pure functions
 */
import { returnAvailablePets } from './availablePets'

/**
 * 1. Lets set up our data inputs to plug in our function in test
 */
const CODES = {
  AVAILABLE: 3
}

const PETS = [

  // Available PET
  {
    name: 'Sammy',
    type: 'dog',
    code: 3,
    vaccinated: true
  },
  {
    name: 'Lemiwinks',
    type: 'cat',
    code: 3,
    vaccinated: true
  },

  // NOT AVAILABLE
  {
    name: 'Aso',
    type: 'dog',
    code: 12,
    vaccinated: true
  },
  {
    name: 'Butch',
    type: 'cat',
    code: 3,
    vaccinated: false
  }
]

/**
 * 2. Beginning of our Spec
 */
describe('Getting Available Pets', () => {

  /**
   * 3. Our Happy Path Test
   */
  it('should return pets that are available and vaccinated', () => {
    expect(returnAvailablePets(PETS, CODES)).toMatchSnapshot()
  })


  /**
   * 4. Our Unhappy path tests
   */
  describe('Not returning available pets', () => {
    it('should not return pets that are not vaccinated and available', () => {
      expect(returnAvailablePets(PETS, CODES)).toMatchSnapshot()
    })
  })

  /**
   * 4. Our Unhappy path tests
   */
  describe('Missing arguments', () => {
    it('should return a false response if pets data structure is not a array', () => {
      expect(returnAvailablePets('not correct type', CODES)).toMatchSnapshot()
    })

    it('should return a false response if pets data is empty', () => {
      expect(returnAvailablePets([], CODES)).toMatchSnapshot()
    })
  })
})