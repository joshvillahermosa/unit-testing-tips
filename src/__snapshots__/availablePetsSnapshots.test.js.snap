// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Getting Available Pets Missing arguments should return a false response if pets data is empty 1`] = `
Object {
  "details": "pets passed in does not have any items",
  "success": false,
}
`;

exports[`Getting Available Pets Missing arguments should return a false response if pets data structure is not a array 1`] = `
Object {
  "details": "pets passed is not type array",
  "success": false,
}
`;

exports[`Getting Available Pets Not returning available pets should not return pets that are not vaccinated and available 1`] = `
Array [
  Object {
    "code": 3,
    "name": "Sammy",
    "type": "dog",
    "vaccinated": true,
  },
  Object {
    "code": 3,
    "name": "Lemiwinks",
    "type": "cat",
    "vaccinated": true,
  },
]
`;

exports[`Getting Available Pets should return pets that are available and vaccinated 1`] = `
Array [
  Object {
    "code": 3,
    "name": "Sammy",
    "type": "dog",
    "vaccinated": true,
  },
  Object {
    "code": 3,
    "name": "Lemiwinks",
    "type": "cat",
    "vaccinated": true,
  },
]
`;
